File: 2020/02/password_philosophy2.py
# password_philosophy.py
# Callum Howard, 2020

from sys import stdin
import re

def solve(line):
    range, letter, password = line.split()
    pos1, pos2 = (int(x) - 1 for x in range.split('-'))
    return sum((password[pos1] == letter[0], password[pos2] == letter[0])) == 1

def solve2(line):
    pos1, pos2, letter, password = re.match(r'(\d+)-(\d+)\s([a-z]):\s(\w+)', line).groups()
    return sum((password[int(pos1) - 1] == letter, password[int(pos2) - 1] == letter)) == 1

print(sum(solve2(line) for line in stdin.readlines()))

File: 2020/02/password_philosophy2.go
// Password Philosophy
// https://adventofcode.com/2020/day/2

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(line string) bool {
	split := strings.Fields(line)
	minMax := strings.Split(split[0], "-")
	pos1, _ := strconv.Atoi(minMax[0])
	pos2, _ := strconv.Atoi(minMax[1])
	letter := split[1][0]
	password := split[2]
	return (password[pos1-1] == letter && password[pos2-1] != letter) ||
		(password[pos1-1] != letter && password[pos2-1] == letter)
}

func main() {
	var scanner = bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanLines)

	var count = 0
	for scanner.Scan() {
		if solve(scanner.Text()) {
			count++
		}
	}
	fmt.Println(count)
}
